doc/src/manual/constructors.md

```diff
@@ -119,23 +119,27 @@ existence by a call to one of the inner constructor methods provided with the ty
 some degree of enforcement of a type's invariants.
 
 If any inner constructor method is defined, no default constructor method is provided: it is presumed
-that you have supplied yourself with all the inner constructors you need. The default constructor
-is equivalent to writing your own inner constructor method that takes all of the object's fields
-as parameters (constrained to be of the correct type, if the corresponding field has a type),
-and passes them to `new`, returning the resulting object:
+that you have supplied yourself with all the inner constructors you need.
+
+The default constructor is equivalent to writing one or two of your own constructor methods that take
+all of the object's fields as arguments.  One is an inner constructor method without type constraints
+on its arguments that passes the them to `new`, returning the resulting object.  The other
+default constructor method is provided if there are any type constraints on the fields. This method
+is an inner constructor method if the type is not generic and similarly passes the arguments to `new`.
```

Comment:
As far as I can tell, the conversion is part of the semantics of the `new` invocation.  That is clear from the fact that the generated code is the same when I write it without conversions.  Assuming that is documented correctly (which I suspect it is not), we don't need to mention the conversions here.  The proposed rewrite omits mention of the `new` invocation; that alone makes your proposed change worse.  It also fails to mention which is an inner/outer constructor, which would be useful guidance to the reader, e.g someone (like me) planning to fix https://github.com/JuliaServices/AutoHashEquals.jl/issues/49.

Reply:
You can see the call to `convert` in the lowered code I posted above. Also:

```julia
julia> struct Bad
           x::Int
       end

julia> Base.convert(::Type{Bad}, x) = error("give me a stacktrace!")

julia> struct Wrapper
           b::Bad
       end

julia> Wrapper(Bad(1))
Wrapper(Bad(1))

julia> Wrapper(1)
ERROR: give me a stacktrace!
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] convert(::Type{Bad}, x::Int64)
   @ Main ./REPL[2]:1
 [3] Wrapper(b::Int64)
   @ Main ./REPL[3]:2
 [4] top-level scope
   @ REPL[5]:1
```

How do you explain this result if we don't actually call `convert`?

Reply:
Sure. But then where does the error come from? I don't see any call to `convert` in your version of the code.

The `convert` gets inserted for you implicitly during lowering, but since it's a Julia function there's no good reason not to be explicit about it when we're describing what Julia actually does. Readers might easily worry that this would perform bit-level reinterpretation, aka
```julia
julia> reinterpret(Float64, Int64(1))
5.0e-324

julia> a = Int64(1)
1

julia> b = reinterpret(Float64, a)
5.0e-324

julia> bitstring(a)
"0000000000000000000000000000000000000000000000000000000000000001"

julia> bitstring(b)
"0000000000000000000000000000000000000000000000000000000000000001"
```
The only reason you can skip it in your implementation is because lowering saves you from disaster by inserting it for you.

The key concept is that the version of the default constructor that doesn't restrict the types of the arguments calls `convert` to give them the types of the fields. It's so important that lowering intervenes and doesn't leave it to chance: you don't get a choice about writing this incorrectly.

I don't understand why you're so resistant to mentioning `convert` when it's actually what's happening.

Reply:
Where does the error come from?  The call to convert is part of the semantics of the `new` operation, and the `new` operation appears in the translation.  Documenting it as part of the semantics of the constructor rather than part of the semantics of `new` would be misleading to the reader, for example someone who intends to perform work like https://github.com/JuliaServices/AutoHashEquals.jl/issues/49 (e.g. me), suggesting to the reader that in order to expand the meaning of the default constructor, one would have to write a conversion in the code.  That is indeed not what one should do.

The `new` operation isn't currently documented sufficiently for that to be clear.  Would it help if I were to add documentation of the semantics of `new`, and mention that the default constructor calls `new` "including the conversions performed by `new`"?